src/pip/_internal/operations/install/wheel.py

```diff
@@ -566,7 +566,7 @@ def is_data_scheme_path(path: RecordPath) -> bool:
         wheel_zip,
         ensure_text(lib_dir, encoding=sys.getfilesystemencoding()),
     )
-    files = map(make_root_scheme_file, root_scheme_paths)
+    files: Iterator[File] = map(make_root_scheme_file, root_scheme_paths)
```

Comment:
Hmm, so Mypy can’t infer this on its own? Maybe we should rewrite this into a generator comprehension instead (which I believe Mypy is able to infer)

Reply:
actually - mypy can infer the type, but it infers a subclass

```
src/pip/_internal/operations/install/wheel.py:585: error: Incompatible types in
assignment (expression has type "chain[File]", variable has type "map[File]")
[assignment]
```

`map` returns a subclass of Iterator `class map`. https://github.com/python/typeshed/tree/master/stdlib#L1062

The later call to chain now returns a `class chain`
https://github.com/python/typeshed/blob/master/stdlib/itertools.pyi#L63
I think thanks to this PR
https://github.com/python/typeshed/pull/5211

So here, were using the typing syntax not to override an uninferred type, but rather to genericize the type of the variable `files`. Unless we explicitly use `typing.cast` Mypy will actually enforce that you type to a superclass - so I think this is actually ok here.

Eg (for illustration):
```
files: int = map(make_root_scheme_file, root_scheme_paths)
```
```
src/pip/_internal/operations/install/wheel.py:569: error: Incompatible types in
assignment (expression has type "map[File]", variable has type "int")
[assignment]
```

It is a bit unergonomic - would be nice if mypy could automatically figure this out. Maybe I can file something upstream with mypy.

Reply:
filed https://github.com/python/mypy/issues/10965

Reply:
That makes sense, thanks. This is a pretty common thing for type inference systems. I don’t think your issue linked above is valid, because the inference logic doea not know whether you indeed want it to automatically upcast, or are explicitly looking for a map object but errornously later assigned a reduce object to the same variable. In this particular example, map does not expose any interface beyond being an iterator, so maybe typeshed should type it as an iterator to begin with (instead of map), but maybe that’d cause unintended side effect as well. Anyway, that’s for Mypy devs to decide. In our case, explicit annotation is the right way to go (and I believe the recommended approach for other type inference systems as well).

Reply:
Actually impressed by `mypy_primer`.

The diff that "broke" this in typeshed ran `mypy_primer` which caught the exact three errors that this deals with:
https://github.com/python/typeshed/pull/5219#issuecomment-819944968

Reply:
Gosh that’s super cool